README.md
Chorder
=======

This repository is essentially a fork of the original SpiffChorder. Some
changes have been made to make the code compile with the latest version
of avr-gcc and the USB driver library has been updated to a recent more
recent version.

The code in this repository will use the Atmega8 as the microcontroller
platform for now.

Original readme:
================

Spaceman Spiff's Chording Keyboard Experiment (SpiffChorder for 
short) is free software; you can redistribute it and/or modify 
it under the terms of the OBDEV license, as found in the 
licence.txt file.
                                                              
SpiffChorder is distributed in the hope that it will be useful,    
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
OBDEV license for further details.
                                                              
Spaceman Spiff's Chording Keyboard Experiment (or SpiffChorder for short)
is an experimental platform for creating a USB chording keyboard. A
chording keyboard or Chorded keyboard is an input device where characters
can be typed by pressing multiple keys at the same time. A popular design 
is the 7-key chorder, where one key is used for each of the index finger, 
middle finger, ring finger and pinky, and 3 keys for the thumb. This can 
be used with either the left or right hand (or a chorder can be used for 
each hand, effectively doubling the typing speed).

For the latest information regarding the SpiffChorder, check the 
project homepage:

  http://symlink.dk/projects/spiffchorder/

and the wiki:

  http://chorder.cs.vassar.edu/spiffchorder/forside

Both the hardware design and software/firmware for the SpiffChorder
is open, and you can make modifications as you see fit. The license is 
based on Objective Development's license, which is now GPL, but with 
the addition that the hardware is documented and open as well.

Objective Development's AVR USB driver is a firmware-only implementation 
of the low speed USB standard (1.5Mbps) which can be used on cheap AVR 
microcontrollers from Atmel.

The hardware for the SpiffChorder is very simple, and consists mainly of 
an ATmega168, a 12MHz crystal, and a few passive components for interfacing 
with the USB port. Additionally, a number of keys are connected for the
chorder. For the current configuration, 7 keys are used, but for future
expansion an eigth input-line has been allocated for the normal chord
decoding, as well as 3 lines that will be mapped directly to keys or
functions (e.g. modifier keys or mouse-clicks for a pointing device).
Optionally 3 LEDs can be connected, intended to show the internal
state/mode of the SpiffChorder.


Legacy hardware
---------------

Because the original design of the SpiffChorder was based on ATmega8
instead of ATmega168, some builders have hardware platforms based on
ATmega8. For this reason it has been decided to try and keep compability
with the old ATmega8 design, with the limitations presented by the
smaller amount of flash memory for program storage. If you are 
constructing the SpiffChorder, use of the ATmega8 is discouraged, because
of code size constraints. One of the nice properties of the SpiffChorder
design is ability to use a USB bootloader, allowing the SpiffChorder
firmware to be updated over the existing USB connection. The bootloader
uses a 2kB bootblock, which leaves only 6kB for code on the ATmega8,
versus 14kB on the ATmega168. At the current stage the code for the
SpiffChorder is very close to the 6kB limit, so just getting a few of 
the planned additional features implemented will likely pass the boundary,
meaning that people with ATmega8-based hardware cannot use the bootloader.

Also, there are plans to extend the SpiffChorder into a composite device,
implementing both a (chording) keyboard and a pointing device, such as the
mini trackball from SparkFun:
http://www.sparkfun.com/commerce/product_info.php?products_id=755

With this addition the code will very likely be larger, even above the
8kB limit, so for this the ATmega168 will probably be the only way to 
go.

If, dispite these warnings, you want to compile the SpiffChorder 
firmware for ATmega8, this should be possible by simply changing the
MCU-line in the Makefile. In the places where hardware differences
influence the code, an attempt has been made to make conditional
code, compiling on the different platforms, based on the MCU selected
in the Makefile.

Since the ATmega8 is not officially supported, especially if changes
are made to the code, pay attention to the size of the code generated.
At the end of the build process you should see a dump of the different
sizes, like the following:
Size after:
main.elf  :
section            size      addr
.text              5904         0
.data                14   8388704
.bss                 76   8388718
.stab              5328         0
....

The number to look for is the size of the .text segment, in this case
5904 bytes. If using the ATmega8, you have 6144 bytes if using a 
bootloader and 8192 bytes otherwise.


Building the hardware
---------------------

The hardware for SpiffChorder is quite simple. Circuit diagrams are
included in the directory "circuit" in PDF and PNG formats. The USB
interface hardware is based on Mindaugas Milasauskas' MJoy application
(http://www.mindaugas.com/projects/MJoy/). Apart from that, the only
interesting things are the 12MHz crystal and interface to the 7 switches
and optionally 3 LEDs.

No circuit board layout is published, mainly because the application
is so simple that it is easier to realize on a perf-board (strip-board),
but the diagram should provide all needed information to create your own
implementation, whether on a perf-board or with printed circuit boards.

A bill of materials (BOM) can be found in circuit/bom.txt, along with some
additional notes regarding the chosen component values.


Building the firmware
---------------------

The SpiffChorder firmware requires avr-gcc and avr-libc (see 
http://www.nongnu.org/avr-libc/ for more details).

Before attempting to compile, edit the Makefile to select the
desired chord mapping and programmer options.

For the easiest approach, you can use WinAVR under Windows, which is
a bundle of the required libraries and tool-chain packaged for
Windows. It even comes with a good editor (Programmers Notepad),
which was used for the development throughout this project.
WinAVR is available from http://winavr.sourceforge.net/.

After installing WinAVR you should be able to run the command
"make" from the directory where the source code is located, which
will build the firmware. The "Makefile" included with the project is
based on an example from the WinAVR distribution, and has support
for burning the firmware to an AVR device through a range of different
programmers supported by the AVRdude programming utility. With a few
modifications to the Makefile it should be possible to accomplish this
programming by issuing the command "make program". Of course these
commands can also be executed from within Programmers Notepad.

Of course you will need some programming hardware to connect between
the circuit and the PC. Several possibilities exist (and are supported
by AVRdude), including some very simple programmers that can be built
on a tight budget. During the experimentation and development of this
project, I have used an STK500 development board from Atmel. If you 
have the possibility of getting one of these, I can highly recommend it,
not only as a programmer, but for a wealth of experiments that can be
done on the board, before any hardware needs to be built.

It is important that the fuses of the microcontroller are set correctly
before the application will work. The most important thing is to configure
the device to use an external crystal. The default configuration of the 
AVR device is to use the internal RC oscillator from the factory, but
this is not sufficiently accurate to handle the delicate timing required 
for USB communication. The Makefile has a target called fuses, 
which uses avrdude to set the fuses for the device (provided the avrdude 
settings in the Makefile have been configured in an appropriate way).

If your programmer does not work with avrdude, you should configure the
fuses as follows (hfuse=0xDD,lfuse=0xFF):
  * Boot flash size = 1024 words
  * Brown-out detection level = 2.7 V
  * External crystal 8.0MHz-xxMHz 16K CK/14 CK + 65ms startup
  * Remember to disable the CKDIV8.

I have had problems when setting the BOD-level to 4.3V, resulting in
Windows saying the device was not recogniced, only to properly detect
it a few seconds later. Setting the BOD-level to 2.7V seems to solve
this.

If using the legacy ATmega8 hardware, the fuses should be configured
as follows: Boot Flash size = 1024 words, BOD-level = 4.0V, BOD enabled,
External crystal, 16K CK + 64ms startup (hfuse=0xD9,lfuse=0x3F).

NOTE: The fuse settings shown here do not enable the bootloader, but
starts the SpiffChorder code directly. If a bootloader is used, you
should set the bootloader enable fuse also (which should be described
in the documentation regarding getting the bootloader in place).

Be aware that the firmware provided here does not use OBDEV's USB 
vendor ID, since this originally was not allowed to be used for HID 
devices. Instead a sepearte VID/PID-pair has been acquired for the
SpiffChorder. You may use this PID freely for SpiffChorders, even
when changing the chord mappings or function of the code.


Files
-----
readme.txt              This file.
changes.txt             What has been changed in the different versions.
doc.txt                 Extra documentation (for further development).
license.txt             The license this project is released under.
circuit/                Directory with circuit diagrams in PDF and PNG formats.
usbdrv/                 Directory with the AVR-USB driver and docs.
keymaps/                Keyboard maps (chord maps).
Makefile                Build instructions for GNU Make.
main.c                  Main source code file.
main.hex                Compiled version, ready to be programmed into uC.
usbconfig.h             Configuration for AVR-USB stack.
                        


License information
-------------------

The SpiffChorder project is based on Objective Development's HIDkeys demo 
application, and although the restriction is mainly that any derivative
works should be published under an Open Source license, the additional
requirements by OBDEV are very resonable, and the entire project has
been released under this license. In other words you should feel free
to use the firmware here for any purpose, but any enhancements should
be given back to the community, in the form of a contribution to OBDEV's
project pages, with full documentation.

See the file license.txt for details.


Copyright
---------

The SpiffChorder project and documentation is Copyright (C) 2006-2008
by Mikkel Holm Olsen (AKA Spaceman Spiff). More information about the
project can be found on http://symlink.dk/projects/spiffchorder/.

$Id: readme.txt 12 2008-03-07 23:20:46Z  $


changes.txt
Changelog for Spaceman Spiff's Chording Keyboard Experiment

This pre-release is for people to test the overall functions, before
making the truely public release of version 1.00.

Other than that, there are a few things that will soon be 
implemented:
1) On/Off-function.
2) LED handling (showing state)
3) Extra modifier keys.

It has not yet been decided if these will be implemented before or after
the release of version 1.00.


Version 0.98 - released 2008-03-08
----------------------------------

 * Bugfix: Repeat no longer generates wrong character at end (======e).
 * Bugfix: Clear modifiers after playing back macro.
 * Bugfix: Clear state after doing extended key.
 * Got proper VID/PID-pair for SpiffChorder.
 * Switched to ATmega168 for more Flash.
 * Changed key tables to 16-bits. Allows modifiers directly in keymaps.


Version 0.97 - released 2008-01-05
----------------------------------

Test release.
 * Additional special strings for * and +.
 * Debounce implemented with hardware timer. Timeout specified in msec.
 * Key repeat timeout using hardware timer. 22msec unit, maximum key repeat
   delay is ~5 sec.
 * Bugfix: leave num/sym-mode after special strings.
 * Implemented multi-mode (one chord for shift+num-mode).


Version 0.92 - not publicly released
------------------------------------

Major overhaul.
 * Chord decoding totally re-written in layered architecture.
 * Implemented "early detect" (key down event when first button released).


Version 0.91 - not publicly released
------------------------------------

Small update.
 * Implemented special strings (multiple characters from single chord).


Version 0.90 - not publicly released
------------------------------------

Initial. Mostly working original feature set.
 * Three modes (normal, function keys, numbers/symbols).
 * switch debouncing with counter.
 * Modifier keys working - modifier locks unused.
 * Boot protocol compatible.

$Id: changes.txt 17 2008-03-08 01:45:31Z  $

keymaps/keycodes.h
/*********************************************************************
 * keycodes.h - Constants used for general keyboard mapping          *
 * $Id:: keycodes.h 16 2008-03-07 23:57:55Z                        $ *
 *********************************************************************
 * SpiffChorder is Copyright (C) 2006-2008 Mikkel Holm Olsen         *
 * based on HID-Test by Christian Starkjohann, Objective Development *
 *********************************************************************
 * Spaceman Spiff's Chording Keyboard Experiment (SpiffChorder for   *
 * short) is free software; you can redistribute it and/or modify    *
 * it under the terms of the OBDEV license, as found in the          *
 * licence.txt file.                                                 *
 *                                                                   *
 * SpiffChorder is distributed in the hope that it will be useful,   *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of    *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the     *
 * OBDEV license for further details.                                *
 *********************************************************************/

#ifndef _KEYCODES_H
#define _KEYCODES_H
#include <avr/pgmspace.h>

#define KEYMAP_USES_MODIFIERS

/* This is the type of the keymap entries */
#ifdef KEYMAP_USES_MODIFIERS
typedef uint16_t keymap_t;
#else
typedef uint8_t keymap_t;
#endif

typedef keymap_t PROGMEM prog_keymap_t;

/*****************************************
 * Possible keycodes in the chord tables *
 *****************************************/
enum keycodes {
/* first part is simply an enumeration of the allowed scan-codes
   used for USB HID devices */
  KEY__=0,              // Unused key-binding
  KEY_errorRollOver,
  KEY_POSTfail,
  KEY_errorUndefined,
  KEY_A,                // 0x04
  KEY_B,
  KEY_C,
  KEY_D,
  KEY_E,
  KEY_F,
  KEY_G,
  KEY_H,
  KEY_I,
  KEY_J,
  KEY_K,
  KEY_L,
  KEY_M,                // 0x10
  KEY_N,
  KEY_O,
  KEY_P,
  KEY_Q,
  KEY_R,
  KEY_S,
  KEY_T,
  KEY_U,
  KEY_V,
  KEY_W,
  KEY_X,
  KEY_Y,
  KEY_Z,
  KEY_1,
  KEY_2,
  KEY_3,                // 0x20
  KEY_4,
  KEY_5,
  KEY_6,
  KEY_7,
  KEY_8,
  KEY_9,
  KEY_0,                // 0x27
  KEY_enter,            // 0x28
  KEY_esc,              // 0x29
  KEY_bckspc,           // 0x2A  backspace
  KEY_tab,              // 0x2B
  KEY_spc,              // 0x2C  space
  KEY_minus,            // 0x2D  - (and _)
  KEY_equal,            // 0x2E  = (and +)
  KEY_lbr,              // 0x2F  [
  KEY_rbr,              // 0x30  ]
  KEY_bckslsh,          // 0x31  \ (and |)
  KEY_hash,             // 0x32  Non-US # and ~
  KEY_smcol,            // 0x33  ; (and :)
  KEY_ping,             // 0x34  ' and "
  KEY_grave,            // 0x35  Grave accent and tilde
  KEY_comma,            // 0x36  , (and <)
  KEY_dot,              // 0x37  . (and >)
  KEY_slash,            // 0x38  / (and ?)
  KEY_cpslck,           // 0x39  capslock
  KEY_F1,
  KEY_F2,
  KEY_F3,
  KEY_F4,
  KEY_F5,
  KEY_F6,
  KEY_F7,               // 0x40
  KEY_F8,
  KEY_F9,
  KEY_F10,
  KEY_F11,
  KEY_F12,
  KEY_PrtScr,
  KEY_scrlck,
  KEY_break,
  KEY_ins,
  KEY_home,
  KEY_pgup,
  KEY_del,
  KEY_end,
  KEY_pgdn,
  KEY_rarr,
  KEY_larr,             // 0x50
  KEY_darr,
  KEY_uarr,
  KEY_numlock,
  KEY_KPslash,
  KEY_KPast,
  KEY_KPminus,
  KEY_KPplus,
  KEY_KPenter,
  KEY_KP1,
  KEY_KP2,
  KEY_KP3,
  KEY_KP4,
  KEY_KP5,
  KEY_KP6,
  KEY_KP7,
  KEY_KP8,              // 0x60
  KEY_KP9,
  KEY_KP0,
  KEY_KPcomma,
  KEY_Euro2,

/* The following are not standard USB HID, but are the modifier keys,
   handeled specially in decoding and mapped to the modifier byte in
   the USB report */
  DIV_Mods,
  MOD_LCTRL=DIV_Mods,   // 0x01
  MOD_LSHIFT,           // 0x02
  MOD_LALT,             // 0x04
  MOD_LGUI,             // 0x08
  MOD_RCTRL,            // 0x10
  MOD_RSHIFT,           // 0x20
  MOD_RALT,             // 0x40
  MOD_RGUI,             // 0x80

/* Next comes the mode change codes */
  DIV_Modes,
  MODE_RESET=DIV_Modes, // Reset (Default mode)
  MODE_MRESET,          // Master Reset (unused)
  MODE_NUM,             // Number/symbols mode
  MODE_NUMLCK,          // Number/symbols lock
  MODE_FUNC,            // Function key mode
  MODE_FUNCLCK,         // Function key lock

/* Then a special mode for both num/sym and shift */
  DIV_Multi,
  MULTI_NumShift=DIV_Multi,

/* And finally macros, that generate multiple key presses */
  DIV_Macro,
  MACRO_000=DIV_Macro,  // 000
  MACRO_00,             // 00
  MACRO_quotes,         // "" and left arrow
  MACRO_parens,         // () and left arrow
  DIV_Last
};

enum modifiers {
  FORCE_LCTRL=0x0100,
  FORCE_LSHIFT=0x0200,
  FORCE_LALT=0x0400,
  FORCE_LGUI=0x0800,
  FORCE_RCTRL=0x1000,
  FORCE_RSHIFT=0x2000,
  FORCE_RALT=0x4000,
  FORCE_RGUI=0x8000,
};


/*********************************************************************************
 * Macro strings defined here - lower bits are key-codes, most significant       *
 * bit is converted to shift.                                                    *
 *********************************************************************************/
const unsigned char MStr_000[] PROGMEM = { KEY_0, KEY_0, KEY_0, 0};
const unsigned char MStr_00[] PROGMEM = { KEY_0, KEY_0, 0};
const unsigned char MStr_quotes[] PROGMEM = { 0x80+KEY_ping, 0x80+KEY_ping, KEY_larr, 0};
const unsigned char MStr_parens[] PROGMEM = { 0x80+KEY_9, 0x80+KEY_0, KEY_larr, 0};

/************************************************************************************
 * Array of the macro strings. Order must be the same as in the keycodes enum       *
 ************************************************************************************/
PROGMEM const unsigned char * const macro_strings[DIV_Last-DIV_Macro] = {
  MStr_000,
  MStr_00,
  MStr_quotes,
  MStr_parens,
};

#endif


main.c
/*********************************************************************
 * main.c - Main firmware                                            *
 * Version 0.98                                                      *
 * $Id:: main.c 17 2008-03-08 01:45:31Z                            $ *
 *********************************************************************
 * SpiffChorder is Copyright (C) 2006-2008 Mikkel Holm Olsen         *
 * based on HID-Test by Christian Starkjohann, Objective Development *
 *********************************************************************
 * Spaceman Spiff's Chording Keyboard Experiment (SpiffChorder for   *
 * short) is free software; you can redistribute it and/or modify    *
 * it under the terms of the OBDEV license, as found in the          *
 * licence.txt file.                                                 *
 *                                                                   *
 * SpiffChorder is distributed in the hope that it will be useful,   *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of    *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the     *
 * OBDEV license for further details.                                *
 *********************************************************************/

#include <avr/io.h>
#include <avr/interrupt.h>
#include <avr/pgmspace.h>
#include <avr/wdt.h>
#include <util/delay.h>
#include <string.h>

#include "usbdrv.h"
#define DEBUG_LEVEL 0
#include "oddebug.h"

/* Hardware documentation:
 * ATmega168 @12.000 MHz
 *
 * PB0     : Pinky key
 * PB1     : Ring finger key
 * PB2     : Middle finger key
 * PB3     : Index finger key
 * PB4     : Near thumb
 * PB5     : Center thumb
 * PB6..PB7: 12MHz X-tal
 * PC0     : Near LED
 * PC1     : Center LED
 * PC2     : Far LED
 * PC3     : Unused modifier 1
 * PC4     : Unused modifier 2
 * PC5     : Unused modifier 3
 * PD0     : D- USB negative (needs appropriate zener-diode and resistors)
 * PD1     : UART TX
 * PD2/INT0: D+ USB positive (needs appropriate zener-diode and resistors)
 * PD6     : Far thumb
 * PD7     : Extra chord button (unused)
 *
 * USB Connector:
 * -------------
 *  1 (red)    +5V
 *  2 (white)  DATA-
 *  3 (green)  DATA+
 *  4 (black)  GND
 *    
 *                                     VCC
 *                  +--[4k7]--+--[2k2]--+
 *      USB        GND        |                     ATmega-8
 *                            |
 *      (D-)-------+----------+--------[82r]------- PD0
 *                 |
 *      (D+)-------|-----+-------------[82r]------- PD2/INT0
 *                 |     |
 *                 _     _
 *                 ^     ^  2 x 3.6V 
 *                 |     |  zener to GND
 *                 |     |
 *                GND   GND
 */

/* LEDs on a boot-protocol keyboard */
#define LED_NUM     0x01
#define LED_CAPS    0x02
#define LED_SCROLL  0x04
#define LED_COMPOSE 0x08
#define LED_KANA    0x10
uchar LEDstate=0;

/* The repeat state and counter in key_repeat can be one of these values
   to signify the state, or it can be higher than REP_STATE_TIMEOUT to
   signify that the timer is running but time-out has not yet occurred */
enum repeat_state_e {
  REP_STATE_READY=0,
  REP_STATE_KEYDOWN_SENT,
  REP_STATE_TIMEOUT
};
/* Key repeat timer/counter. if>REP_STATE_TIMEOUT then decreased by main loop */
volatile uchar key_repeat=REP_STATE_READY;

uchar repeat_value=REP_STATE_TIMEOUT+23;    // 12M/(2^18)/23 = 502ms timeout
uchar debounce_value=10;  // 10 ms debounce. Should be loaded from EEPROM.
uchar use_early_detect=1; // This is a configuration option

/* Originally used as a mask for the modifier bits, but now also
   used for other x -> 2^x conversions (lookup table). */
const char modmask[8] PROGMEM = {
  0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80
};


/* USB report descriptor (length is defined in usbconfig.h)
   This has been changed to conform to the USB keyboard boot
   protocol */
const char usbHidReportDescriptor[USB_CFG_HID_REPORT_DESCRIPTOR_LENGTH] 
  PROGMEM = {
    0x05, 0x01,                    // USAGE_PAGE (Generic Desktop)
    0x09, 0x06,                    // USAGE (Keyboard)
    0xa1, 0x01,                    // COLLECTION (Application)
    0x05, 0x07,                    //   USAGE_PAGE (Keyboard)
    0x19, 0xe0,                    //   USAGE_MINIMUM (Keyboard LeftControl)
    0x29, 0xe7,                    //   USAGE_MAXIMUM (Keyboard Right GUI)
    0x15, 0x00,                    //   LOGICAL_MINIMUM (0)
    0x25, 0x01,                    //   LOGICAL_MAXIMUM (1)
    0x75, 0x01,                    //   REPORT_SIZE (1)
    0x95, 0x08,                    //   REPORT_COUNT (8)
    0x81, 0x02,                    //   INPUT (Data,Var,Abs)
    0x95, 0x01,                    //   REPORT_COUNT (1)
    0x75, 0x08,                    //   REPORT_SIZE (8)
    0x81, 0x03,                    //   INPUT (Cnst,Var,Abs)
    0x95, 0x05,                    //   REPORT_COUNT (5)
    0x75, 0x01,                    //   REPORT_SIZE (1)
    0x05, 0x08,                    //   USAGE_PAGE (LEDs)
    0x19, 0x01,                    //   USAGE_MINIMUM (Num Lock)
    0x29, 0x05,                    //   USAGE_MAXIMUM (Kana)
    0x91, 0x02,                    //   OUTPUT (Data,Var,Abs)
    0x95, 0x01,                    //   REPORT_COUNT (1)
    0x75, 0x03,                    //   REPORT_SIZE (3)
    0x91, 0x03,                    //   OUTPUT (Cnst,Var,Abs)
    0x95, 0x06,                    //   REPORT_COUNT (6)
    0x75, 0x08,                    //   REPORT_SIZE (8)
    0x15, 0x00,                    //   LOGICAL_MINIMUM (0)
    0x25, 0x65,                    //   LOGICAL_MAXIMUM (101)
    0x05, 0x07,                    //   USAGE_PAGE (Keyboard)
    0x19, 0x00,                    //   USAGE_MINIMUM (Reserved (no event indicated))
    0x29, 0x65,                    //   USAGE_MAXIMUM (Keyboard Application)
    0x81, 0x00,                    //   INPUT (Data,Ary,Abs)
    0xc0                           // END_COLLECTION  
};

/* An array of pointers to the decoding tables */
const prog_keymap_t * const keymap_tables[] PROGMEM = {
  keymap_default,                  // The default mode
  keymap_numsym,                   // Numbers/Symbols mode
  keymap_function                  // Function key mode
};

const prog_keymap_t *current_keymap = keymap_default;
const prog_keymap_t *special_keymap = NULL;

/* The ReportBuffer contains the USB report sent to the PC */
static uchar reportBuffer[8];    /* buffer for HID reports */
static uchar idleRate;           /* in 4 ms units */
static uchar protocolVer=1;      /* 0 is the boot protocol, 1 is report protocol */

/* Debounce counter. if>1 then decreased by timer interrupt */
volatile uchar debounce=0;

static void hardwareInit(void) {
  PORTB = 0x3F;   /* Enable pull-ups */
  DDRB  = 0x00;   /* Port B are switch inputs */

  PORTC = 0x38;   /* LEDs on PC0..PC2, switches on PC3..PC5 */
  DDRC  = 0x07;   /* PC0..PC2 are outputs, PC3..PC5 are inputs */
  
  PORTD = 0xfa;   /* 1111 1010 bin: activate pull-ups except on USB lines */
  DDRD  = 0x07;   /* 0000 0111 bin: all pins input except USB (-> USB reset) */

  /* USB Reset by device only required on Watchdog Reset */
  _delay_us(11);  /* delay >10ms for USB reset */ 

  DDRD = 0x02;    /* 0000 0010 bin: remove USB reset condition */

#if defined (__AVR_ATmega168__)
# define TIFR_REG TIFR0 
# define CMA_VECT TIMER2_COMPA_vect
  /* configure timer 0 for a rate of 12M/(1024 * 256) = 45.78 Hz (~22ms) */
  TCCR0A = 0;     /* Normal Mode, OC disabled */
  TCCR0B = 5;     /* timer 0 prescaler: 1024 */
  /* configure timer 2 for a rate of 12M/(256 * 47) = 997.3 Hz (~1ms) */
  OCR2A = 47;
  TCCR2A = (1<<WGM21); /* CTC-mode */
  TCCR2B = (6<<CS20);  /* prescaler=256 */
  TIMSK2|=(1<<OCIE2A);  /* Enable compare match interrupt */
#elif defined (__AVR_ATmega8__)
# define TIFR_REG TIFR
# define CMA_VECT TIMER2_COMP_vect
  /* configure timer 0 for a rate of 12M/(1024 * 256) = 45.78 Hz (~22ms) */
  TCCR0 = 5;           /* timer 0 prescaler: 1024 */
  /* configure timer 2 for a rate of 12M/(256 * 47) = 997.3 Hz (~1ms) */
  OCR2 = 47;
  TCCR2 = (1<<WGM21)|(6<<CS20); /* CTC-mode, prescaler=256 */
  TIMSK|=(1<<OCIE2);   /* Enable compare match interrupt */
#else
#error Selected microcontroller not supported
#endif

}

/* Timer 2 interrupt handler - decrease debounce counter every 1ms */
ISR(CMA_VECT) {
  /* This gets called every 1ms and updates the button debounce counter */
  if (debounce>1) {
    --debounce;
  }
}

/* The modifier keys applied to next keypress and the modifier locks */
uchar mod_state=0, mod_lock=0;

/* The current mode and whether to lock the mode or only apply it to one keypress */
uchar current_mode=0, lock_mode=0, last_modifiers=0;

/* Used to send the key_up event */
const uint8_t PROGMEM key_up[] = {0};

/* This pointer tells us something needs to be sent before returning to keyboard scanning */
const uint8_t *pendingPtr = key_up;

/* Handle a chord */
unsigned char process_chord(uchar chord) {
  uint8_t key;
#ifdef KEYMAP_USES_MODIFIERS
  keymap_t keymods;
  keymods=pgm_read_word(special_keymap?&special_keymap[chord&0x7F]:
                    &current_keymap[chord&0x7F]);
  key=(uint8_t)keymods; /* typecast masks off upper bits */
#else
  key=pgm_read_byte(special_keymap?&special_keymap[chord&0x7F]:
                    &current_keymap[chord&0x7F]);
#endif
  special_keymap=NULL;
  if (key>=DIV_Mods) { // This is not a standard key
    if (key>=DIV_Multi) { // Either Multi-mode or Macro
      if (key>=DIV_Macro) { /* Macro */
        pendingPtr=(const uint8_t*)pgm_read_word(&macro_strings[key-DIV_Macro]);
        mod_state&=mod_lock; 
        if (!lock_mode) { // Reset mode unless it was locked
          current_keymap=keymap_default;
          current_mode=0;
        }
      } else { // Multi
        special_keymap=keymap_numsym; // Numbers/symbols-mode
        mod_state|=0x02; // Left shift
      }
    } else { // Either mods or modes
      if (key>=DIV_Modes) { /* Mode change requested */
        current_mode=(key-DIV_Modes);
        lock_mode=current_mode&0x01;
        current_mode>>=1;
        current_keymap=(const prog_keymap_t*)pgm_read_word(&keymap_tables[current_mode]);
      } else { // Modifier
        mod_state|=pgm_read_byte(&modmask[key-DIV_Mods]);
      }
    }
    key=0;
  }

  if (key) {
    memset(reportBuffer,0,sizeof(reportBuffer)); /* Clear report buffer */
    reportBuffer[2]=key;

#ifdef KEYMAP_USES_MODIFIERS
    key=keymods>>8;
    if (key) {
      reportBuffer[0]=key;
    } else 
#endif
    {
      reportBuffer[0]=mod_state;
    }
    last_modifiers=reportBuffer[0]; /* Save the modifiers for proper clean-up on release */

    mod_state&=mod_lock;
    if (!lock_mode) {
      current_keymap=keymap_default;
      current_mode=0;
    }
    return 1;
  }
  return 0;
}

/* When the state of keys change, handle key-down or key-up */
uchar key_change(uchar data, uchar timeout) {
  static uchar latched=0xFF, adding=0;
  uchar retval=0;
  if (timeout) {
    // If the same chord has been held for an amount of time - send key-down - repeat starts
    if (adding) retval=process_chord(~latched);
  } else {
    if (latched & ~data) { // buttons(s) pressed
      adding=1;
      latched&=data;
      key_repeat=repeat_value;
    } else if (~latched & data) { // buttons(s) released
      if (use_early_detect?adding:data==0xFF) {
        if (key_repeat!=REP_STATE_KEYDOWN_SENT) retval=process_chord(~latched);
        latched=data;
        adding=0;
      }
      if ((retval)||(key_repeat==REP_STATE_KEYDOWN_SENT)) { // Some key was reported
        if (!pendingPtr) pendingPtr=key_up; // Make sure the key up event is sent immediately
      }
      key_repeat=REP_STATE_READY;
      if (use_early_detect) latched=data;
    }
  }
  return retval;
}


/* Scan the keys, debounce, and report changes in state */
uchar scankeys(void) {
  static uchar olddata=0xFF;
  uchar retval=0;
  uchar data;

  data=(PINB&0x3F)|(PIND&0xC0); // Read keys
  if (olddata^data) { // Change detected
    if (debounce==0) {
      debounce=debounce_value;
    } else if (debounce==1) {
      retval=key_change(data,0); // Process key change
      olddata=data;
      debounce=0;
    }
  } else {
    if (key_repeat==REP_STATE_TIMEOUT) {
      retval=key_change(data,1); // Send key-down
      key_repeat=REP_STATE_KEYDOWN_SENT;
    }
  }

  return retval;
}

unsigned char getCharFromPending(void) {
  static char state=0, kval=0;
  memset(reportBuffer,0,sizeof(reportBuffer)); /* Clear report buffer */
  if (state) {
    reportBuffer[0]=last_modifiers;
    state=0;
  } else {
    kval=pgm_read_byte(pendingPtr);
    reportBuffer[0]=(kval&0x80?0x02:0);
    reportBuffer[2]=kval&~0x80;
    if (kval) {
      pendingPtr++;
    } else {
      pendingPtr=NULL;
    }
    state=1;
  }
  return 1;
}

uchar expectReport=0;

uchar usbFunctionSetup(uchar data[8]) {
  usbRequest_t *rq = (void *)data;
  usbMsgPtr = reportBuffer;
  if((rq->bmRequestType & USBRQ_TYPE_MASK) == USBRQ_TYPE_CLASS){    /* class request type */
    if(rq->bRequest == USBRQ_HID_GET_REPORT){  
      /* wValue: ReportType (highbyte), ReportID (lowbyte) */
      /* we only have one report type, so don't look at wValue */
      return sizeof(reportBuffer);
    }else if(rq->bRequest == USBRQ_HID_SET_REPORT){
      if (rq->wLength.word == 1) { /* We expect one byte reports */
        expectReport=1;
        return 0xFF; /* Call usbFunctionWrite with data */
      }  
    }else if(rq->bRequest == USBRQ_HID_GET_IDLE){
      usbMsgPtr = &idleRate;
      return 1;
    }else if(rq->bRequest == USBRQ_HID_SET_IDLE){
      idleRate = rq->wValue.bytes[1];
    }else if(rq->bRequest == USBRQ_HID_GET_PROTOCOL) {
      if (rq->wValue.bytes[1] < 1) {
        protocolVer = rq->wValue.bytes[1];
      }
    }else if(rq->bRequest == USBRQ_HID_SET_PROTOCOL) {
      usbMsgPtr = &protocolVer;
      return 1;
    }
  }
  return 0;
}

/* The write function is called when LEDs should be set */
uchar usbFunctionWrite(uchar *data, uchar len) {
  if ((expectReport)&&(len==1)) {
    LEDstate=data[0]; /* Get the state of all 5 LEDs */
  }
  expectReport=0;
  return 0x01;
}

int main(void) {
  uchar   updateNeeded = 0;
  uchar   idleCounter = 0;

  wdt_enable(WDTO_2S); /* Enable watchdog timer 2s */
  hardwareInit(); /* Initialize hardware (I/O) */
  
  odDebugInit();

  usbInit(); /* Initialize USB stack processing */
  sei(); /* Enable global interrupts */
  
  for(;;){  /* Main loop */
    wdt_reset(); /* Reset the watchdog */
    usbPoll(); /* Poll the USB stack */
    
    if (!updateNeeded) {
      if (pendingPtr) {
        updateNeeded=getCharFromPending(); /* Keys pending in output buffer? */
      } else {
        updateNeeded=scankeys(); /* Scan the keyboard for changes */
      }
    }

    /* Check timer if we need periodic reports */
    if (TIFR_REG & (1<<TOV0)) {
      TIFR_REG = 1<<TOV0; /* Reset flag */
      if (key_repeat>REP_STATE_TIMEOUT) {
        --key_repeat;
      }
      if (idleRate != 0) { /* Do we need periodic reports? */
        if(idleCounter > 4){ /* Yes, but not yet */
          idleCounter -= 5;   /* 22 ms/tick */
        } else { /* Yes, it is time now */
          updateNeeded = 1;
          idleCounter = idleRate;
        }
      }
    }
    
    /* If an update is needed, send the report */
    if (updateNeeded && usbInterruptIsReady()) {
      updateNeeded = 0;
      usbSetInterrupt(reportBuffer, sizeof(reportBuffer));
    }
  }
  return 0;
}
